Add capability to move parameters from one object to another

GiudGiud opened the pull request:
The way to build these new Physics, see https://github.com/idaholab/moose/issues/25642, is going to pull params from people's existing actions & objects

I think we want to avoid:

- re-writing docstring for every parameters in the coming Physics action(s)
- re-setting the same defaults etc

Ideally we would even move the parameter checking (dependences etc) to the validParam as well so we could transfer that too

For example, in this new ground breaking heat conduction physics:
```
InputParameters
HeatConductionPhysics::validParams()
{
  InputParameters params = PhysicsBase::validParams();
  params.addClassDescription("Add the heat conduction physics");

  params.transferParam<MaterialPropertyName>(ADHeatConduction::validParams(),
                                             "thermal_conductivity");
```

GiudGiud committed:
Add capability to transfer a parameter from one object to another
refs #25642

GiudGiud committed:
Add unit test for transferring parameters

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1803096/) on 362ff73 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25648/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1803114/) on 362ff73 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4e9d07</th>
      <th colspan=4>#25648 362ff7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.69%</td>
      <td>85.70%</td>
      <td>+0.01%</td>
      <td>93.48%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91796</td>
      <td>91851</td>
      <td>+55</td>
      <td>43</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15331</td>
      <td>15322</td>
      <td>-9</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25648/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25648/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25648/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Strongify hasDefault logic, still not perfect

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -939,6 +948,18 @@ class InputParameters : public Parameters
+   */
+  template <typename T>
+  void transferParam(const InputParameters & source_param,
+                     std::string name,
+                     std::string new_description = "");
```
```suggestion
                     const std::string & name,
                     const std::string & new_description = "");
```

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+template <typename T>
+void
+InputParameters::transferParam(const InputParameters & source_params,
+                               std::string name_in,
+                               std::string new_description)
```
```suggestion
                               const std::string & name_in,
                               const std::string & new_description)
```

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+        name_in,
+        "' parameter could not be transferred because it does not exist in the source parameters");
+
+  std::string description =
```
```suggestion
  const std::string description =
```

lindsayad reviewed (changes requested):
An easy way to "work-around" and achieve the concept you are implementing here is to create global functions or other static methods that return common `InputParameters`. I do this quite regularly. I didn't mind the `transferParam` concept but then I saw the length of the method

GiudGiud commented:
> An easy way to "work-around" and achieve the concept you are implementing here is to create global functions or other static methods that return common InputParameters.

You need to create somewhat arbitrary groups if you're just trying to get one parameter from a kernel. I started like this, I've seen this in your PRs

> then I saw the length of the method

That's kind of a consequence of us having almost a dozen types of templated parameters. All useful, but not all sharing the same abstractions

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+
+  if (source_params.isParamRequired(name))
+  {
+    // Use the name_in that was specified in the transfer, not the potentially deprecated name
+    // Check for a variable parameter
```
`name` should be the name that is guaranteed to be non-deprecated. `name_in` may be deprecated.

GiudGiud replied:
but name_in is the name the developer passed when writing transferParam

I think it s confusing to be renaming that one to the name that was deprecated on the other object

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+    // Use the name_in that was specified in the transfer, not the potentially deprecated name
+    // Check for a variable parameter
+    if (source_params.hasCoupledValue(name))
+      addRequiredCoupledVar(name, description);
```
```suggestion
      addRequiredCoupledVar(name_in, description);
```
? This would be consistent with your other code

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+    if (source_params.hasCoupledValue(name))
+    {
+      if (!source_params.hasDefaultCoupledValue(name))
+        addCoupledVar(name, description);
+      else if (source_params.numberDefaultCoupledValues(name) == 1)
+        addCoupledVar(name, source_params.defaultCoupledValue(name), description);
+      else
+      {
+        std::vector<Real> coupled_values;
+        for (const auto i : make_range(source_params.numberDefaultCoupledValues(name)))
+          coupled_values.push_back(source_params.defaultCoupledValue(name, i));
+        addCoupledVar(name, coupled_values, description);
```
When adding the parameter, in most places you are using `name_in`, but here you are using `name`. If there is a special treatment for coupled variables vs. other parameters, it would be good to explain it

GiudGiud replied:
Ok I ll take another look

GiudGiud replied:
there is not. there is only bad copy pasting

GiudGiud commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+    if (source_params.hasCoupledValue(name))
+    {
+      if (!source_params.hasDefaultCoupledValue(name))
+        addCoupledVar(name, description);
```
```suggestion
        addCoupledVar(name_in, description);
```

GiudGiud commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+      if (!source_params.hasDefaultCoupledValue(name))
+        addCoupledVar(name, description);
+      else if (source_params.numberDefaultCoupledValues(name) == 1)
+        addCoupledVar(name, source_params.defaultCoupledValue(name), description);
```
```suggestion
        addCoupledVar(name_in, source_params.defaultCoupledValue(name), description);
```

GiudGiud commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+        std::vector<Real> coupled_values;
+        for (const auto i : make_range(source_params.numberDefaultCoupledValues(name)))
+          coupled_values.push_back(source_params.defaultCoupledValue(name, i));
+        addCoupledVar(name, coupled_values, description);
```
```suggestion
        addCoupledVar(name_in, coupled_values, description);
```

GiudGiud commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,79 @@ InputParameters::have_parameter(std::string_view name_in) const
+        addRangeCheckedParam<T>(name_in, source_params.rangeCheckedFunction(name), description);
+    }
+    else if constexpr (std::is_same_v<MooseEnum, T> || std::is_same_v<MultiMooseEnum, T>)
+      addParam<T>(name_in, source_params.get<T>(name), description);
```
```suggestion
      addParam<T>(name_in, source_params.get<T>(name_in), description);
```

GiudGiud committed:
Address Alex s review

Be more consistent about the name of the parameter to transfer
const references when possible to aovid copies
Add a warning if using a deprecated name to transfer a parameter

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
> name_in is the parameter name passed to the API sure 

and that s why it needs to be the final name
I can add a warning

lindsayad commented:
> and that s why it needs to be the final name

I never said it shouldn't be. My complaint was with the inaccurate comment

lindsayad commented on framework/include/utils/InputParameters.h:
```diff
@@ -1830,3 +1851,82 @@ InputParameters::have_parameter(std::string_view name_in) const
+
+  if (source_params.isParamRequired(name))
+  {
+    // Use the name_in that was specified in the transfer, not the potentially deprecated name
```
```suggestion
    // Use the name_in that was specified in the transfer
```

GiudGiud commented:
Thanks for the review! Sorry for misunderstanding it
